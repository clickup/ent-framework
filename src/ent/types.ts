import type { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  ID,
  LoadByInput,
  Order,
  Table,
  UniqueKey,
  Where,
} from "../types";
import type { Validation } from "./Validation";
import type { VC } from "./VC";

/**
 * A very shallow interface of Ent class (as a collection of static methods).
 * Used in some places where we need the very minimum from the Ent.
 */
export interface EntClass<TTable extends Table = any> {
  readonly SCHEMA: Schema<TTable>;
  readonly VALIDATION: Validation<TTable>;
  readonly name: string; // class constructor name

  new (): Ent;
  loadX(vc: VC, id: string): Promise<Ent>;
  loadNullable(vc: VC, id: string): Promise<Ent | null>;
  loadIfReadableNullable(vc: VC, id: string): Promise<Ent | null>;
  count(vc: VC, where: CountInput<TTable>): Promise<number>;
  exists(vc: VC, where: ExistsInput<TTable>): Promise<boolean>;
  select(
    vc: VC,
    where: Where<TTable>,
    limit: number,
    order?: Order<TTable>
  ): Promise<Ent[]>;
  selectChunked(
    vc: VC,
    where: Where<TTable>,
    chunkSize: number,
    limit: number,
    custom?: {}
  ): AsyncIterableIterator<Ent[]>;
  loadByX(vc: VC, keys: LoadByInput<TTable, UniqueKey<TTable>>): Promise<Ent>;
}

/**
 * A very shallow interface of one Ent.
 */
export interface Ent {
  readonly [ID]: string;
  readonly vc: VC;
  deleteOriginal(): Promise<boolean>;
}
