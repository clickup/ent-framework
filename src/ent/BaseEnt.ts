import type { Client } from "../abstract/Client";
import type { Cluster } from "../abstract/Cluster";
import type { Schema } from "../abstract/Schema";
import type { Table, UniqueKey } from "../types";
import { CacheMixin } from "./mixins/CacheMixin";
import { ConfigMixin } from "./mixins/ConfigMixin";
import type { HelpersClass } from "./mixins/HelpersMixin";
import { HelpersMixin } from "./mixins/HelpersMixin";
import { PrimitiveMixin } from "./mixins/PrimitiveMixin";

/**
 * This is a helper function to create new Ent classes. Run once per each
 * Ent+Cluster on app boot. See examples in __tests__/TestObjects.ts and
 * EntTest.ts.
 *
 * Since all Ent objects are immutable (following the modern practices),
 * 1. Ent is not a DataMapper pattern;
 * 2. Ent is not an ActiveRecord;
 * 3. At last, Ent is not quite a DAO pattern too.
 *
 * We assume that Ents are very simple (we don't need triggers or multi-ent
 * touching mutations), because we anyway have a GraphQL layer on top of it.
 *
 * Finally, a naming decision has been made: we translate database field names
 * directly to Ent field names, no camelCase. This has proven its simplicity
 * benefits in the past: the less translation layers we have, the easier it is
 * to debug and develop.
 */
export function BaseEnt<
  TTable extends Table,
  TUniqueKey extends UniqueKey<TTable>,
  TClient extends Client,
>(
  cluster: Cluster<TClient>,
  schema: Schema<TTable, TUniqueKey>,
): HelpersClass<TTable, TUniqueKey, TClient> {
  return HelpersMixin(
    CacheMixin(PrimitiveMixin(ConfigMixin(class {}, cluster, schema))),
  );
}
