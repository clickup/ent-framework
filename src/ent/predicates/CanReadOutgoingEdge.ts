import { EntClass } from "../types";
import { VC } from "../VC";
import { IDsCacheReadable, Predicate } from "./Predicate";

/**
 * Checks that an ent which a field is pointing to is readable:
 *
 * EntOur[company_id] ---> EntCompany[id]
 *
 * This predicate delegates the readability permission check for the current ent
 * to another ent with ID equals to the value of our ent's field.
 *
 * - field = user_id in the above example
 * - toEntClass = EntCompany in the above example
 */

export class CanReadOutgoingEdge<TField extends string>
  implements Predicate<Record<TField, string | null>>
{
  readonly name = this.constructor.name + "(" + this.field + ")";

  constructor(
    public readonly field: TField,
    public readonly toEntClass: EntClass
  ) {}

  async check(vc: VC, row: Record<TField, string | null>): Promise<boolean> {
    const toID = row[this.field];
    if (!toID) {
      return false;
    }

    const cache = vc.cache(IDsCacheReadable);
    if (cache.has(toID!)) {
      return true;
    }

    await this.toEntClass.loadX(vc, toID!);
    // sill here and not thrown? save to the cache
    cache.add(toID!);
    return true;
  }
}
