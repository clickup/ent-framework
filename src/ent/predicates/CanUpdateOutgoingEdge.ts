import { EntClass } from "../types";
import { VC } from "../VC";
import { IDsCacheUpdatable, Predicate } from "./Predicate";

/**
 * Checks that an Ent available via a field is updatable. See
 * CanReadOutgoingEdge comments for more details.
 */
export class CanUpdateOutgoingEdge<TField extends string>
  implements Predicate<Record<TField, string | null>>
{
  readonly name = this.constructor.name + "(" + this.field + ")";

  constructor(
    public readonly field: TField,
    public readonly toEntClass: EntClass
  ) {}

  async check(vc: VC, row: Record<TField, string | null>): Promise<boolean> {
    const toID = row[this.field];
    if (!toID) {
      return false;
    }

    const cache = vc.cache(IDsCacheUpdatable);
    if (cache.has(toID!)) {
      return true;
    }

    // load the target Ent and check that it's updatable
    const toEnt = await this.toEntClass.loadX(vc, toID!);
    await this.toEntClass.VALIDATION.validateUpdate(
      vc,
      toEnt as any,
      {},
      true /* privacyOnly */
    );

    // sill here and not thrown? save to the cache
    cache.add(toID!);
    return true;
  }
}
