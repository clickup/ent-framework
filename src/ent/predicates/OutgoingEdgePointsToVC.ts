import type { VC } from "../VC";
import type { Predicate } from "./Predicate";

/**
 * Checks that the field's value is the same as VC's principal:
 *
 * EntOur[user_id] ---> vc.principal
 */
export class OutgoingEdgePointsToVC<TField extends string>
  implements Predicate<Record<TField, string | null>>
{
  readonly name;

  constructor(public readonly field: TField) {
    this.name = this.constructor.name + "(" + this.field + ")";
  }

  async check(vc: VC, row: Record<TField, string | null>): Promise<boolean> {
    const toID = row[this.field];
    if (!toID) {
      return false;
    }

    return toID === vc.principal;
  }
}
