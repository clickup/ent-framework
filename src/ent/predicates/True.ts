import { VC } from "../VC";
import { Predicate } from "./Predicate";

/**
 * Always passes; used for e.g. globally accessed objects.
 */
export class True implements Predicate<never> {
  readonly name = this.constructor.name;

  async check(_vc: VC, _row: object): Promise<boolean> {
    return true;
  }
}
