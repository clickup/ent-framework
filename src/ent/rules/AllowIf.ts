import { EntAccessError } from "../errors/EntAccessError";
import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule } from "./Rule";

/**
 * Returns ALLOW if the predicate succeeds, otherwise SKIP.
 * - Used mostly for read permission checks.
 * - This rule may still throw an exception if the exception is a wild one (not
 *   derived from EntAccessError).
 */
export class AllowIf<TInput extends object> extends Rule<TInput> {
  readonly _TAG!: "AllowIf";

  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    try {
      return (await this.predicate.check(vc, input))
        ? { decision: "ALLOW", rule: this, cause: null }
        : { decision: "SKIP", rule: this, cause: null };
    } catch (error: unknown) {
      if (error instanceof EntAccessError) {
        // We carry a cause for this SKIP decision too if it was due to an
        // access-related error.
        return { decision: "SKIP", rule: this, cause: error };
      }

      throw error;
    }
  }
}
