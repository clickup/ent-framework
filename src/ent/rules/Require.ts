import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule, RuleDecision } from "./Rule";

/**
 * Returns TOLERATE if the predicate succeeds, otherwise DENY.
 * - Used mostly for write permission checks.
 * - This rule may still throw an exception if it's a wild one (i.e. not derived
 *   from EntAccessError).
 */
export class Require<TInput extends object> extends Rule<TInput> {
  readonly _TAG!: "Require";

  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    return (await this.predicate.check(vc, input))
      ? { decision: RuleDecision.TOLERATE, rule: this, cause: null }
      : { decision: RuleDecision.DENY, rule: this, cause: null };
  }
}
