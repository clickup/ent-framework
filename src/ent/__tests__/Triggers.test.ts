import { inspect } from "util";
import {
  recreateTestTables,
  testCluster,
} from "../../sql/__tests__/test-utils";
import { SQLSchema } from "../../sql/SQLSchema";
import { ID } from "../../types";
import { BaseEnt, GLOBAL_SHARD } from "../BaseEnt";
import { CanReadOutgoingEdge } from "../predicates/CanReadOutgoingEdge";
import { CanUpdateOutgoingEdge } from "../predicates/CanUpdateOutgoingEdge";
import { OutgoingEdgePointsToVC } from "../predicates/OutgoingEdgePointsToVC";
import { True } from "../predicates/True";
import { AllowIf } from "../rules/AllowIf";
import { Require } from "../rules/Require";
import type { VC } from "../VC";
import { createVC, expectToMatchSnapshot } from "./test-utils";

const $EPHEMERAL = Symbol("$EPHEMERAL");
const $EPHEMERAL2 = Symbol("$EPHEMERAL2");

/**
 * User
 */
export class EntTestUser extends BaseEnt(
  testCluster,
  new SQLSchema(
    'ent.triggers"user',
    {
      id: { type: ID, autoInsert: "id_gen()" },
      name: { type: String },
      url_name: { type: String, allowNull: true },
      is_alseeing: { type: Boolean, autoInsert: "false" },
      created_at: { type: Date, autoInsert: "now()" },
      updated_at: { type: Date, autoUpdate: "now()" },
    },
    ["name"]
  )
) {
  static readonly CREATE = [
    `CREATE TABLE %T(
      id bigint NOT NULL PRIMARY KEY,
      company_id bigint DEFAULT NULL,
      name text NOT NULL,
      url_name text,
      is_alseeing boolean,
      created_at timestamptz NOT NULL,
      updated_at timestamptz NOT NULL,
      UNIQUE (name)
    )`,
  ];

  static override configure() {
    return new this.Configuration({
      shardAffinity: GLOBAL_SHARD,
      privacyInferPrincipal: async (_vc, { id }) => id,
      privacyLoad: [new AllowIf(new OutgoingEdgePointsToVC("id"))],
      privacyInsert: [],
      privacyUpdate: [new Require(new OutgoingEdgePointsToVC("id"))],
    });
  }

  nameUpper(): string {
    return this.name.toUpperCase();
  }
}

/**
 * Headline -> User -> Company
 */
export class EntTestHeadline extends BaseEnt(
  testCluster,
  new SQLSchema(
    'ent.triggers"headline',
    {
      id: { type: ID, autoInsert: "id_gen()" },
      user_id: { type: ID },
      headline: { type: String },
      name: { type: String, allowNull: true, autoInsert: "NULL" },
      [$EPHEMERAL]: { type: String, allowNull: true }, // required, but nullable
      [$EPHEMERAL2]: { type: Number, autoInsert: "NULL" }, // optional (can be skipped), but if present, must be non-nullable
    },
    []
  )
) {
  static readonly CREATE = [
    `CREATE TABLE %T(
      id bigint NOT NULL PRIMARY KEY,
      user_id bigint NOT NULL,
      headline text NOT NULL,
      name text
    )`,
  ];

  static readonly TRIGGER_CALLS: Array<{
    type: string;
    old?: any;
    new?: any;
    input?: any;
  }> = [];

  static override configure() {
    return new this.Configuration({
      shardAffinity: [],
      privacyLoad: [
        new AllowIf(new CanReadOutgoingEdge("user_id", EntTestUser)),
      ],
      privacyInsert: [
        new Require(new CanUpdateOutgoingEdge("user_id", EntTestUser)),
      ],
      beforeInsert: [
        async (_vc, { input }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "beforeInsert1",
            input: { ...input },
          });
          input.headline += " added-by-beforeInsert1";
          if (input[$EPHEMERAL2]) {
            input[$EPHEMERAL2]! += 1000;
          }
        },
        async (_vc, { input }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "beforeInsert2",
            input: { ...input },
          });
          input.headline += " added-by-beforeInsert2";
        },
      ],
      beforeUpdate: [
        async (_vc, { newRow, oldRow, input }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "beforeUpdate",
            old: oldRow,
            new: newRow,
            input: { ...input },
          });
          input.headline = newRow.headline + " added-by-beforeUpdate";
          if (input[$EPHEMERAL2]) {
            input[$EPHEMERAL2]! += 1000000;
          }
        },
      ],
      beforeDelete: [
        async (_vc, { oldRow }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "beforeDelete",
            old: oldRow,
          });
        },
      ],
      afterInsert: [
        async (_vc, { input }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "afterInsert",
            input: { ...input },
          });
        },
      ],
      afterUpdate: [
        async (_vc, { newRow, oldRow }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "afterUpdate",
            old: oldRow,
            new: newRow,
          });
        },
        [
          (_vc, row) => JSON.stringify([row.name]),
          async (_vc, { newRow, oldRow }) => {
            EntTestHeadline.TRIGGER_CALLS.push({
              type: "afterUpdate (if name changed)",
              old: oldRow,
              new: newRow,
            });
          },
        ],
      ],
      afterDelete: [
        async (_vc, { oldRow }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "afterDelete",
            old: oldRow,
          });
        },
      ],
      afterMutation: [
        async (_vc, { newOrOldRow }) => {
          EntTestHeadline.TRIGGER_CALLS.push({
            type: "afterMutation",
            input: newOrOldRow,
          });
        },
        [
          (_vc, row) => JSON.stringify([row.name]),
          async (_vc, { newOrOldRow }) => {
            EntTestHeadline.TRIGGER_CALLS.push({
              type: "afterMutation (if name changed)",
              input: newOrOldRow,
            });
          },
        ],
      ],
    });
  }
}

/**
 * Country
 */
export class EntTestCountry extends BaseEnt(
  testCluster,
  new SQLSchema(
    'ent.triggers"country',
    {
      id: { type: String, autoInsert: "id_gen()" },
      name: { type: String, allowNull: true, autoInsert: "NULL" },
    },
    ["name"]
  )
) {
  static readonly CREATE = [
    `CREATE TABLE %T(
      id bigint NOT NULL PRIMARY KEY,
      name text,
      UNIQUE(name)
    )`,
  ];

  static readonly TRIGGER_CALLS: Array<{
    type: string;
    old?: any;
    new?: any;
    input?: any;
  }> = [];

  static override configure() {
    return new this.Configuration({
      shardAffinity: GLOBAL_SHARD,
      privacyLoad: [new AllowIf(new True())],
      privacyInsert: [new Require(new True())],
      beforeInsert: [
        async (_vc, { input }) => {
          this.TRIGGER_CALLS.push({
            type: "beforeInsert",
            input: { ...input },
          });
        },
      ],
      beforeUpdate: [
        async (_vc, { newRow, oldRow, input }) => {
          this.TRIGGER_CALLS.push({
            type: "beforeUpdate",
            old: oldRow,
            new: newRow,
            input: { ...input },
          });
        },
      ],
      beforeDelete: [
        async (_vc, { oldRow }) => {
          this.TRIGGER_CALLS.push({
            type: "beforeDelete",
            old: oldRow,
          });
        },
      ],
      afterInsert: [
        async (_vc, { input }) => {
          this.TRIGGER_CALLS.push({
            type: "afterInsert",
            input: { ...input },
          });
        },
      ],
      afterUpdate: [
        async (_vc, { newRow, oldRow }) => {
          this.TRIGGER_CALLS.push({
            type: "afterUpdate",
            old: oldRow,
            new: newRow,
          });
        },
      ],
      afterDelete: [
        async (_vc, { oldRow }) => {
          this.TRIGGER_CALLS.push({
            type: "afterDelete",
            old: oldRow,
          });
        },
      ],
      afterMutation: [
        async (_vc, { newOrOldRow }) => {
          this.TRIGGER_CALLS.push({
            type: "afterMutation",
            input: newOrOldRow,
          });
        },
      ],
    });
  }
}

let vc: VC;

beforeEach(async () => {
  await recreateTestTables([EntTestUser, EntTestHeadline, EntTestCountry]);

  const user = await EntTestUser.insertReturning(createVC().toOmniDangerous(), {
    name: "John",
    url_name: "john",
  });
  expect(user.vc.principal).toEqual(user.id);
  vc = user.vc;
});

test("triggers", async () => {
  EntTestHeadline.TRIGGER_CALLS.splice(0);
  await EntTestHeadline.insertReturning(vc, {
    user_id: vc.principal,
    headline: "xyz",
    [$EPHEMERAL]: null,
  });
  expectToMatchSnapshot(
    inspect(EntTestHeadline.TRIGGER_CALLS),
    "0: insert happened"
  );

  EntTestHeadline.TRIGGER_CALLS.splice(0);
  const headline = await EntTestHeadline.insertReturning(vc, {
    user_id: vc.principal,
    headline: "abc",
    [$EPHEMERAL]: "eph",
    [$EPHEMERAL2]: 42,
  });
  expect(headline.headline).toEqual(
    "abc added-by-beforeInsert1 added-by-beforeInsert2"
  );
  expectToMatchSnapshot(
    inspect(EntTestHeadline.TRIGGER_CALLS),
    "1: insert happened"
  );

  EntTestHeadline.TRIGGER_CALLS.splice(0);
  const headline2 = await headline.updateReturningX({
    headline: "xyz-updated",
    name: "new-name",
    [$EPHEMERAL2]: 101,
  });
  expect(headline2.headline).toEqual("xyz-updated added-by-beforeUpdate");
  expectToMatchSnapshot(
    inspect(EntTestHeadline.TRIGGER_CALLS),
    "2: update happened"
  );

  EntTestHeadline.TRIGGER_CALLS.splice(0);
  const headline3 = await headline.updateReturningX({
    [$EPHEMERAL]: "eph3",
  });
  expect(headline3.headline).toEqual(
    "abc added-by-beforeInsert1 added-by-beforeInsert2 added-by-beforeUpdate"
  );
  expectToMatchSnapshot(
    inspect(EntTestHeadline.TRIGGER_CALLS),
    "3: noop-update happened"
  );

  EntTestHeadline.TRIGGER_CALLS.splice(0);
  await headline.updateChanged({
    [$EPHEMERAL]: "eph4",
  });
  expectToMatchSnapshot(
    inspect(EntTestHeadline.TRIGGER_CALLS),
    "4: updateChanged happened"
  );

  EntTestHeadline.TRIGGER_CALLS.splice(0);
  await headline.deleteOriginal();
  expect(await EntTestHeadline.loadNullable(vc, headline.id)).toBeNull();
  expectToMatchSnapshot(
    inspect(EntTestHeadline.TRIGGER_CALLS),
    "5: delete happened"
  );
});

test("skip after triggers if operation soft fails", async () => {
  await EntTestCountry.insertReturning(vc, { name: "zzz" });

  EntTestCountry.TRIGGER_CALLS.splice(0);
  const abc = await EntTestCountry.insertReturning(vc, { name: "abc" });
  expectToMatchSnapshot(
    inspect(EntTestCountry.TRIGGER_CALLS),
    "1: insert happened"
  );

  EntTestCountry.TRIGGER_CALLS.splice(0);
  await EntTestCountry.insertIfNotExists(vc, { name: "abc" });
  expectToMatchSnapshot(
    inspect(EntTestCountry.TRIGGER_CALLS),
    "2: insert soft-failed on unique key conflict"
  );

  await abc.deleteOriginal();

  EntTestCountry.TRIGGER_CALLS.splice(0);
  await abc.updateOriginal({ name: "zzz" });
  expectToMatchSnapshot(
    inspect(EntTestCountry.TRIGGER_CALLS),
    "3: update soft-failed on non-existing row"
  );

  EntTestCountry.TRIGGER_CALLS.splice(0);
  await abc.deleteOriginal();
  expectToMatchSnapshot(
    inspect(EntTestCountry.TRIGGER_CALLS),
    "4: delete soft-failed on non-existing row"
  );
});

test("inserting Ents with custom IDs when they have beforeInsert triggers", async () => {
  const ent = await EntTestCountry.insertReturning(vc.toOmniDangerous(), {
    name: "Test",
  });
  await ent.deleteOriginal();
  const newEnt = await EntTestCountry.insertReturning(
    vc.toOmniDangerous(),
    ent
  );
  expect(newEnt.id).toEqual(ent.id);
});
