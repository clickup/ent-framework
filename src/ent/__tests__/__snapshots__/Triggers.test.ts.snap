// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skip after triggers if operation soft fails: 1: insert happened 1`] = `
"[
  {
    type: 'beforeInsert',
    input: { name: 'abc', id: '<id>' }
  },
  {
    type: 'afterInsert',
    input: { name: 'abc', id: '<id>' }
  },
  {
    type: 'afterMutation',
    input: { name: 'abc', id: '<id>' }
  }
]"
`;

exports[`skip after triggers if operation soft fails: 2: insert soft-failed on unique key conflict 1`] = `
"[
  {
    type: 'beforeInsert',
    input: { name: 'abc', id: '<id>' }
  }
]"
`;

exports[`skip after triggers if operation soft fails: 3: update soft-failed on non-existing row 1`] = `
"[
  {
    type: 'beforeUpdate',
    old: EntTestCountry { id: '<id>', name: 'abc' },
    new: { id: '<id>', name: 'zzz' },
    input: { name: 'zzz' }
  }
]"
`;

exports[`skip after triggers if operation soft fails: 4: delete soft-failed on non-existing row 1`] = `
"[
  {
    type: 'beforeDelete',
    old: EntTestCountry { id: '<id>', name: 'abc' }
  }
]"
`;

exports[`triggers: 0: insert happened 1`] = `
"[
  {
    type: 'beforeInsert1',
    input: {
      user_id: '<id>',
      headline: 'xyz',
      id: '<id>',
      [Symbol($EPHEMERAL)]: null
    }
  },
  {
    type: 'beforeInsert2',
    input: {
      user_id: '<id>',
      headline: 'xyz added-by-beforeInsert1',
      id: '<id>',
      [Symbol($EPHEMERAL)]: null
    }
  },
  {
    type: 'afterInsert',
    input: {
      user_id: '<id>',
      headline: 'xyz added-by-beforeInsert1 added-by-beforeInsert2',
      id: '<id>',
      [Symbol($EPHEMERAL)]: null
    }
  },
  {
    type: 'afterMutation',
    input: {
      user_id: '<id>',
      headline: 'xyz added-by-beforeInsert1 added-by-beforeInsert2',
      id: '<id>',
      [Symbol($EPHEMERAL)]: null
    }
  },
  {
    type: 'afterMutation (if name changed)',
    input: {
      user_id: '<id>',
      headline: 'xyz added-by-beforeInsert1 added-by-beforeInsert2',
      id: '<id>',
      [Symbol($EPHEMERAL)]: null
    }
  }
]"
`;

exports[`triggers: 1: insert happened 1`] = `
"[
  {
    type: 'beforeInsert1',
    input: {
      user_id: '<id>',
      headline: 'abc',
      id: '<id>',
      [Symbol($EPHEMERAL)]: 'eph',
      [Symbol($EPHEMERAL2)]: 42
    }
  },
  {
    type: 'beforeInsert2',
    input: {
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1',
      id: '<id>',
      [Symbol($EPHEMERAL)]: 'eph',
      [Symbol($EPHEMERAL2)]: 1042
    }
  },
  {
    type: 'afterInsert',
    input: {
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      id: '<id>',
      [Symbol($EPHEMERAL)]: 'eph',
      [Symbol($EPHEMERAL2)]: 1042
    }
  },
  {
    type: 'afterMutation',
    input: {
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      id: '<id>',
      [Symbol($EPHEMERAL)]: 'eph',
      [Symbol($EPHEMERAL2)]: 1042
    }
  },
  {
    type: 'afterMutation (if name changed)',
    input: {
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      id: '<id>',
      [Symbol($EPHEMERAL)]: 'eph',
      [Symbol($EPHEMERAL2)]: 1042
    }
  }
]"
`;

exports[`triggers: 2: update happened 1`] = `
"[
  {
    type: 'beforeUpdate',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    },
    new: {
      id: '<id>',
      user_id: '<id>',
      headline: 'xyz-updated',
      name: 'new-name',
      [Symbol($EPHEMERAL2)]: 101
    },
    input: {
      headline: 'xyz-updated',
      name: 'new-name',
      [Symbol($EPHEMERAL2)]: 101
    }
  },
  {
    type: 'afterUpdate',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    },
    new: {
      id: '<id>',
      user_id: '<id>',
      headline: 'xyz-updated added-by-beforeUpdate',
      name: 'new-name',
      [Symbol($EPHEMERAL2)]: 1000101
    }
  },
  {
    type: 'afterUpdate (if name changed)',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    },
    new: {
      id: '<id>',
      user_id: '<id>',
      headline: 'xyz-updated added-by-beforeUpdate',
      name: 'new-name',
      [Symbol($EPHEMERAL2)]: 1000101
    }
  },
  {
    type: 'afterMutation',
    input: {
      id: '<id>',
      user_id: '<id>',
      headline: 'xyz-updated added-by-beforeUpdate',
      name: 'new-name',
      [Symbol($EPHEMERAL2)]: 1000101
    }
  },
  {
    type: 'afterMutation (if name changed)',
    input: {
      id: '<id>',
      user_id: '<id>',
      headline: 'xyz-updated added-by-beforeUpdate',
      name: 'new-name',
      [Symbol($EPHEMERAL2)]: 1000101
    }
  }
]"
`;

exports[`triggers: 3: noop-update happened 1`] = `
"[
  {
    type: 'beforeUpdate',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    },
    new: {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null,
      [Symbol($EPHEMERAL)]: 'eph3'
    },
    input: { [Symbol($EPHEMERAL)]: 'eph3' }
  },
  {
    type: 'afterUpdate',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    },
    new: {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2 added-by-beforeUpdate',
      name: null,
      [Symbol($EPHEMERAL)]: 'eph3'
    }
  },
  {
    type: 'afterMutation',
    input: {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2 added-by-beforeUpdate',
      name: null,
      [Symbol($EPHEMERAL)]: 'eph3'
    }
  }
]"
`;

exports[`triggers: 4: updateChanged happened 1`] = `
"[
  {
    type: 'beforeUpdate',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    },
    new: {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null,
      [Symbol($EPHEMERAL)]: 'eph4'
    },
    input: { [Symbol($EPHEMERAL)]: 'eph4' }
  },
  {
    type: 'afterUpdate',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    },
    new: {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2 added-by-beforeUpdate',
      name: null,
      [Symbol($EPHEMERAL)]: 'eph4'
    }
  },
  {
    type: 'afterMutation',
    input: {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2 added-by-beforeUpdate',
      name: null,
      [Symbol($EPHEMERAL)]: 'eph4'
    }
  }
]"
`;

exports[`triggers: 5: delete happened 1`] = `
"[
  {
    type: 'beforeDelete',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    }
  },
  {
    type: 'afterDelete',
    old: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    }
  },
  {
    type: 'afterMutation',
    input: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    }
  },
  {
    type: 'afterMutation (if name changed)',
    input: EntTestHeadline {
      id: '<id>',
      user_id: '<id>',
      headline: 'abc added-by-beforeInsert1 added-by-beforeInsert2',
      name: null
    }
  }
]"
`;
